lib/iris/_data_manager.py

```diff
@@ -0,0 +1,471 @@
+        assert state, emsg.format(self._realised_dtype)
+
+        # Ensure validity of lazy data with realised dtype.
+        state = not (not self.has_lazy_data() and
+                     self._realised_dtype is not None)
+        emsg = ('Unexpected real data with realised dtype, got '
+                'real data and realised {!r}.')
+        assert state, emsg.format(self._realised_dtype)
+
+    def _deepcopy(self, memo, data=None, realised_dtype=None):
+        """
+        Perform a deepcopy of the :class:`~iris._data_manager.DataManager`
+        instance.
+
+        Args:
+
+        * memo:
+            :class:`copy` memo dictionary.
+
+        Kwargs:
+
+        * data:
+            Replacement data to substitute the currently managed
+            data with.
+
+        * realised_dtype:
+            Replacement for the intended dtype of the realised lazy data.
+
+        Returns:
+            :class:`~iris._data_manager.DataManager` instance.
+
+        """
+        try:
+            if data is None:
+                # Copy the managed data.
+                if self.has_lazy_data():
+                    data = copy.deepcopy(self._lazy_array, memo)
+                else:
+                    data = self._real_array.copy()
+            else:
+                # Check that the replacement data is valid relative to
+                # the currently managed data.
+                DataManager(self.core_data).replace(data)
+                # If the replacement data is valid, then use it but
+                # without copying it.
+
+            result = DataManager(data, realised_dtype=realised_dtype)
+        except ValueError as error:
+            emsg = 'Cannot copy {!r} - {}'
+            raise ValueError(emsg.format(type(self).__name__, error))
+
+        return result
+
+    def _realised_dtype_setter(self, realised_dtype):
+        """
+        Set the intended dtype of the realised lazy data. This is to support
+        the case of lazy masked integral and boolean data in dask.
+
+        Args:
+
+        * realised_dtype:
+            A numpy :class:`~numpy.dtype`, array-protocol type string,
+            or built-in scalar type.
+
+        """
+        if realised_dtype is None:
+            self._realised_dtype = None
+        else:
+            if not self.has_lazy_data():
+                emsg = 'Cannot set realised dtype, no lazy data is available.'
+                raise ValueError(emsg)
+            realised_dtype = np.dtype(realised_dtype)
+            if realised_dtype.kind not in 'biu':
+                emsg = ('Can only cast lazy data to an integer or boolean '
+                        'dtype, got {!r}.')
+                raise ValueError(emsg.format(realised_dtype))
+            self._realised_dtype = realised_dtype
+
+            # Check the manager contract, as the managed dtype has changed.
+            self._assert_axioms()
+
+    @property
+    def core_data(self):
```

Comment:
As a note for our future selves, I still like the idea of calling this `data`. Then we can rename `data` to `real_data` and keep `lazy_data` as `lazy_data`. I think this is the most descriptive set of names for these variables.

But let's not do that in this PR! One thing at a time.